src/utils/entityStore.ts

```diff
@@ -3,95 +3,90 @@
-   * @param {(entity: T) => boolean} [filter] optional method that is called while
-   * searching for the entity nearest a point. If the filter returns false, the point
-   * is considered invalid and is not considered. If the filter returns true, the point
-   * is considered valid and will be considered.
+   * @param {(entity: T) => boolean} [filter] optional method that is called
+   * while searching for the entity nearest a point. If the filter returns
+   * false, the point is considered invalid and is not considered. If the filter
+   * returns true, the point is considered valid and will be considered.
    * @returns {T} Will return the nearest entity or undefined if none are found
    */
   entityNearest(point: Point, filter?: (entity: T) => boolean): T;
-  /**
-   * Iterator that loops through entities and returns a transformed array
-   * @param {(value: T) => S} [callback] transformation function that is passed
-   * passed an entity {T} and returns an object {S}.
-   * @returns {S[]} The aggregate result of each call to the transformation function
-   */
-  map<S>(callback: (value: T) => S): S[];
 
   /**
-   * Iterator that loops through entities calls the callback on each iteration.
-   * @param {(value: T) => void} [callback] transformation function that is passed
-   * passed an entity {T}.
+   * Returns the current internal array of entities.
+   *
+   * @returns {T[]} the current internal array of entities.
    */
-  forEach(callback: (value: T) => void): void;
-}
-
-export interface IPositionedEntity {
-  position: Point;
+  entities(): T[];
 }
 
 /**
- * Array-backed implementation of {EntityStore}
+ * Implementation of {IEntityStore} that uses an array for easy iteration as
+ * well as a quad tree for fast nearest-point queries.
+ *
+ * Note that if the position of your entities changes, you MUST rebuild the
+ * entity store for the `entityNearest` method to work since the quadtree does
+ * not know that its nodes have moved.
  */
-export class EntityArray<T extends IPositionedEntity> implements IEntityStore<T> {
+export class EntityStore<T extends IPositionedEntity> implements IEntityStore<T> {
   private _entities: T[];
+  private _tree: d3.Quadtree<T>;
 
   constructor() {
     this._entities = [];
+    this._tree = d3.quadtree<T>()
+      // int casting for faster computation. losing sub-pixel precision here is
+      // of no concern.
+      .x((d) => d.position.x | 0)
+      .y((d) => d.position.y | 0);
   }
 
-  public add(entity: T) {
-    this._entities.push(entity);
+  public addAll(entities: T[]) {
+    this._entities.push(...entities);
+    this._tree.addAll(entities);
   }
 
   /**
-   * Iterates through array of of entities and computes the closest point using
-   * the standard Euclidean distance formula.
+   * Looks up the nearest data point using the internal quadtree.
+   *
+   * Note that for performance reasons, the optional filter is **DESTRUCTIVE**
```

Comment:
It does, which is why i made this comment. There are no guarantees about who will call entityNearest from the outside, but unfortunately creating a copy of the tree every time is quite expensive. What would be ideal is if we could just get the quadtree itself to do the filtering, but that's not how it's implemented